### PR TITLE
[RLlib] SAC/DQN activate multi-agent learning tests and small bug fix in `MultiAgentEpisode`.

### PR DESCRIPTION
**Pull request by sven1977**
- SAC/DQN activates multi-agent learning tests for CartPole (DQN) and Pendulum (SAC)
- small bug fix in `MultiAgentEpisode.concat()`: self.env_t_to_agent_t is not properly built in resulting episode object.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/BUILD [hunk 1]
```diff
@@ -303,6 +292,15 @@ py_test(
     args = ["--as-test", "--enable-new-api-stack"]
 )
 
+py_test(
```
Comment (simonsays1980) [inline rllib/BUILD:295]: Yey! And we are there! :)

File: rllib/BUILD [hunk 2]
```diff
@@ -452,6 +450,15 @@ py_test(
...
```
Comment (simonsays1980) [inline rllib/BUILD:453]: Next step: synchronized sampling.

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -3413,6 +3413,20 @@ def _compile_iteration_results_new_api_stack(
...
+        #  keys. Otherwise, we'll have to guess here, what "popular" keys users use in
+        #  order to protect them from running into Tune KeyErrors.
+        if ENV_RUNNER_RESULTS not in results:
+            results[ENV_RUNNER_RESULTS] = {}
+        for must_have in [
+            EPISODE_RETURN_MEAN,
+            EPISODE_RETURN_MIN,
+            EPISODE_RETURN_MAX,
+        ]:
+            if must_have not in results[ENV_RUNNER_RESULTS]:
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm.py:3427]: Yup, understood. But this means, if a user collects custom metrics via callback and uses Tune on this ... an error could result due to this metrics not being available in an iteration. We should document this somewhere - maybe at the MetricsLogger API ...

File: rllib/env/multi_agent_episode.py [hunk 4]
```diff
@@ -873,7 +873,12 @@ def concat_episode(self, other: "MultiAgentEpisode") -> None:
                     )
 
                 # Concatenate the env- to agent-timestep mappings.
-                self.env_t_to_agent_t[agent_id].extend(other.env_t_to_agent_t[agent_id])
+                j = self.env_t
+                for i, val in enumerate(other.env_t_to_agent_t[agent_id][1:]):
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:877]: This is one of the nasty parts of the MAE. Awesome catch!

File: rllib/tuned_examples/dqn/cartpole_dqn.py [hunk 5]
```diff
@@ -15,16 +15,11 @@
...
-    .api_stack(
-        enable_rl_module_and_learner=True,
-        enable_env_runner_and_connector_v2=True,
-    )
     .rl_module(
         # Settings identical to old stack.
         model_config_dict={
             "fcnet_hiddens": [256],
-            "fcnet_activation": "relu",
+            "fcnet_activation": "tanh",
```
Comment (simonsays1980) [inline rllib/tuned_examples/dqn/cartpole_dqn.py:22]: beasty

File: rllib/tuned_examples/sac/multi_agent_pendulum_sac.py [hunk 6]
```diff
@@ -55,13 +37,23 @@
         },
         num_steps_sampled_before_learning_starts=256,
     )
+    .rl_module(
+        model_config_dict={
+            "fcnet_hiddens": [256, 256],
+            "fcnet_activation": "relu",
+            # "post_fcnet_hiddens": [],
```
Comment (simonsays1980) [inline rllib/tuned_examples/sac/multi_agent_pendulum_sac.py:44]: This part is old stack equivalent.

File: rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py [hunk 7]
```diff
@@ -131,7 +131,7 @@ def add(
         """
         episodes: List["MultiAgentEpisode"] = force_list(episodes)
 
-        new_episode_ids: List[str] = {eps.id_ for eps in episodes}
+        new_episode_ids: Set[str] = {eps.id_ for eps in episodes}
```
Comment (simonsays1980) [inline rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py:134]: Great catch!

File: rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py [hunk 8]
```diff
@@ -828,57 +828,52 @@ def _update_module_counters(self, ma_episode: MultiAgentEpisode) -> None:
             # Only add if the agent has stepped in the episode (chunk).
             if agent_steps > 0:
                 # Receive the corresponding module ID.
-                # TODO (sven, simon): Is there always a mapping? What if not?
-                # Is then module_id == agent_id?
-                module_id = ma_episode._agent_to_module_mapping[agent_id]
+                module_id = ma_episode.module_for(agent_id)
```
Comment (simonsays1980) [inline rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py:831]: Sweet, this method is just clean!

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome work! Off-policy now learns!

File: rllib/algorithms/algorithm.py (hunk 3 above)
Comment (sven1977) [reply rllib/algorithms/algorithm.py:3427]: Correct, that's why I put the comment there. We should actually fix Tune, there is no other solution to this problem, imo.

Tune didn't probably think of this b/c in SL, you don't have "strangely behaving" episodes that don't deliver data sometimes.

File: rllib/tuned_examples/sac/multi_agent_pendulum_sac.py (hunk 6 above)
Comment (sven1977) [reply rllib/tuned_examples/sac/multi_agent_pendulum_sac.py:44]: got it!

**Pull request merged**